### ISSUE TITLE
Event Handling - Tutorial (Example 1: Bouncing Ball)

### ISSUE THREAD

**Issue by anirudhpammi** [SciML/DifferentialEquations.jl#570]
I tried the code given on the link (https://docs.juliadiffeq.org/latest/features/callback_functions/). The Tsit5() adaptive algorithm doesn't work for this example. Can you please tell me why that is the case? When I disable the adaptive step size, specify the dt=1e-3 and solve this example, I get matching results. 

Unrelated question: I have a different case, I need to check if one of the state variables of my system have reached a certain value. If I use the DiscreteCallback, I notice that there is an overshoot in my simulations. Should I be using the ContinuousCallback for better results?

**Comment by ChrisRackauckas** [comment]
>I tried the code given on the link (https://docs.juliadiffeq.org/latest/features/callback_functions/). The Tsit5() adaptive algorithm doesn't work for this example. Can you please tell me why that is the case? When I disable the adaptive step size, specify the dt=1e-3 and solve this example, I get matching results.

Which example? Maybe @kanav99 could take a look.

>If I use the DiscreteCallback, I notice that there is an overshoot in my simulations. Should I be using the ContinuousCallback for better results?

They do very different things. DiscreteCallbacks act at the end of a step, ContinuousCallback acts at a rootfinding condition. If you need to hit a specific value, like the bouncing ball, use the latter. If you want to do something like plot after each step, use the former.

**Comment by kanav99** [comment]
Hi! I just tried the example on OrdinaryDiffEq@master and DiffEqBase@master and it works for me. Did you make any changes to the code to cause error? Also are you on updated DifferentialEquations.jl?

**Comment by anirudhpammi** [comment]
@ChrisRackauckas : Thank you! That makes it clear. 
@kanav99 : It is the example of the bouncing ball, I am using the v6.9.0 of DifferentialEquations. The only change I made was to use a different plotting backend. I use the PyPlot backend and pygui option to  plot the figures in an external window. If I use the Plots package and plot the solution using the command "plot(sol)" it seems to work fine. I guess it means that the plot command makes sense of the solution object and plots it in a meaningful way. Do you know if there is a way around this so that I can use just the sol.t and sol.u points to plot a meaningful answer without any backend magic? Thank you.

**Comment by anirudhpammi** [comment]
@ChrisRackauckas : Just one more follow up question, for the continuouscallback, what sort of root finding method do you use? Thanks a lot!

**Comment by kanav99** [comment]
That seems to be some plotting issue then. Let me check it.
We use Bisection as of now for rootfinding.

**Comment by ChrisRackauckas** [comment]
You have to make sure you plot the discontinuity points on both sides to resolve it correctly. What does your plot look like? Sounds like you just plotted a standard plot over time, so you got a straight line between each point.

**Comment by anirudhpammi** [comment]
@ChrisRackauckas : Yes, that is exactly what I was doing and I get a straight line. I will look into it . I will read the documentation about the discontinuity points.

@kanav99 : Thank you, I seem to have a problem with the algorithm, at some time step my state variable steps exactly on the point which is a zero of the function it is supposed to minimize, and the bisection method fails to identity it as a solution. The documentation says that if the starting condition is < 10*eps() (by default) the root can't be found and since my starting point is exactly the root and event is not triggered. Is this intended? Could you suggest a workaround? Thanks!

**Comment by ChrisRackauckas** [comment]
Sounds like it was just a user plotting issue. Please feel free to discuss more if there's more to it!

**Comment by anirudhpammi** [comment]
@ChrisRackauckas : Sorry for the delay in replying. I did have a second question about the root finding algorithm: at some time step my state variable steps exactly on the point which is a zero of the function it is supposed to minimize, and the bisection method fails to identity it as a solution. The documentation says that if the starting condition is < 10*eps() (by default) the root can't be found and since my starting point is exactly the root and event is not triggered. Is this intended? Could you suggest a workaround? Thanks!

**Comment by ChrisRackauckas** [comment]
The start is treated differently, since otherwise a lot of callbacks would have weird side effects. For that reason we put in an `initialize!` option in the callback interface. If you want the effect to apply at the starting point, then you can add `initialize = integrator -> affect!(integrator)`. By default it's a function that does nothing.

**Comment by anirudhpammi** [comment]
My starting point isn't the problem. The state variable exactly steps on the value that minimizes the function during the simulation. Since the step is exactly on the value, the root finding algorithm fails. I encountered a similar problem with a python package as well. Is that normal?

**Comment by ChrisRackauckas** [comment]
Hmm, that is odd. Can you share an example of that? We can probably find a way to fix it.

**Comment by anirudhpammi** [comment]
So the idea is as follows: 
You have two nodes $X_1$ and $X_2$

**Comment by anirudhpammi** [comment]
So the idea is as follows: 
You have two nodes $X_1$ and $X_2$

$e^{i \pi} = -1$

**Comment by anirudhpammi** [comment]
The brief description of the problem is in the pdf:
[Solver_stepping_exactly_on_root (1).pdf](https://github.com/SciML/DifferentialEquations.jl/files/4399278/Solver_stepping_exactly_on_root.1.pdf)

```julia
using PyPlot, DifferentialEquations, BenchmarkTools
pygui(true)
close("all")

# Parameters
g = -0.4
α = 1
t_exc_1 = []
t_exc_2 = []


function Es(t,t_exc)
    t_redef = (-t_exc .+ t) .* Int.(t .>= t_exc)
    return ( length(t_redef) == 0 ? 0 : sum((g*α^2) .* exp.(-α .* t_redef) .* t_redef)  )
end

function neuron_activation!(du,u,p,t)
    # du -> rate of change, u -> state variable/initial conditions, p -> parameters, t -> time
    X = p[1]
    du[1] = X - u[1] + Es(t,t_exc_2)
    du[2] = X - u[2] + Es(t,t_exc_1)
end

function condition(u,t,integrator)
    (u[1]-1)
end

function affect!(integrator)
    integrator.u[1] = 0
    push!(t_exc_1,integrator.t)
end

function condition2(u,t,integrator)
    (u[2] - 1)
end

function affect2!(integrator)
    integrator.u[2] = 0
    push!(t_exc_2,integrator.t)
end

cb1 = ContinuousCallback(condition,affect!)
cb2 = ContinuousCallback(condition2,affect2!)

u0 = [0.0;0.001]
tspan = (0.0,200)
X_thresh = 1.3
p = [X_thresh]
prob = ODEProblem(neuron_activation!,u0,tspan,p);
alg = Vern7();
sol = solve(prob,alg,callback=CallbackSet(cb2,cb1),reltol=1e-6,abstol=1e-7)

figure()
plot(sol.t,sol.u)
T_p = diff(t_exc_1)[end]

```

**Comment by anirudhpammi** [comment]
@ChrisRackauckas : Did you have a chance to look at it? Thanks!

**Comment by ChrisRackauckas** [comment]
I don't understand the issue. I ran:

```julia
using DifferentialEquations, BenchmarkTools

# Parameters
g = -0.4
α = 1
t_exc_1 = []
t_exc_2 = []


function Es(t,t_exc)
    t_redef = (-t_exc .+ t) .* Int.(t .>= t_exc)
    return ( length(t_redef) == 0 ? 0 : sum((g*α^2) .* exp.(-α .* t_redef) .* t_redef)  )
end

function neuron_activation!(du,u,p,t)
    # du -> rate of change, u -> state variable/initial conditions, p -> parameters, t -> time
    X = p[1]
    du[1] = X - u[1] + Es(t,t_exc_2)
    du[2] = X - u[2] + Es(t,t_exc_1)
end

function condition(u,t,integrator)
    (u[1]-1)
end

function affect!(integrator)
    integrator.u[1] = 0
    @show integrator.t
    push!(t_exc_1,integrator.t)
end

function condition2(u,t,integrator)
    (u[2] - 1)
end

function affect2!(integrator)
    integrator.u[2] = 0
    @show integrator.t
    push!(t_exc_2,integrator.t)
end

cb1 = ContinuousCallback(condition,affect!)
cb2 = ContinuousCallback(condition2,affect2!)

u0 = [0.0;0.001]
tspan = (0.0,200)
X_thresh = 1.3
p = [X_thresh]
prob = ODEProblem(neuron_activation!,u0,tspan,p);
alg = Vern7();
sol = solve(prob,alg,callback=CallbackSet(cb2,cb1),reltol=1e-6,abstol=1e-7)

using Plots
Plots.plot(sol,plotdensity=10000)
T_p = diff(t_exc_1)[end]
```

and it seems both are hit around every 4 seconds?

**Comment by anirudhpammi** [comment]
Could you please try with u0 = [0.0;0.0] instead of u0=[0.0,0.001]. 
I notice that with the default value of reltol and abstol, it seems to work fine, but with the values I gave (reltol=1e-6,abstol=1e-7), the solutions doesn't work correctly. 
![Figure_1](https://user-images.githubusercontent.com/6874193/82656704-7f7ee600-9c24-11ea-8f45-b94918dbcd2f.png)

**Comment by kanav99** [comment]
@anirudhpammi  does this plot look good? I think this issue is fixed due to recent changes.
<img width="1005" alt="Screenshot 2020-12-11 at 18 51 05" src="https://user-images.githubusercontent.com/33966400/101908276-d577a300-3be1-11eb-836a-2927ecb21eb1.png">

**Comment by ChrisRackauckas** [comment]
That looks correct to me.

**Comment by anirudhpammi** [comment]
Yes, it looks correct. thanks!
